### PR TITLE
fixed pca for features > samples, and fixed pca_noise_estimate

### PR DESCRIPTION
**Pull request by samcoveney**
Following on from https://github.com/dipy/dipy/discussions/2678 I have attempted to do several things:

1. fix `genpca` to work when the number of samples (voxels) is less than the number of features (images) by modifying `_pca_classifier`
2. fix `pca_noise_estimate` to be able to use the canonical algorithm

This is a new pull request to replace #2707 which became totally corrupted for reasons I don't understand, and I simply could not fix. Sorry about that. But I addressed all comments left at the time.

### CODE REVIEW

**Comment by samcoveney** [comment]
Hi, I opened this around 3 weeks ago, with the original corrupted pull on 4 Jan, will it be possible to assign someone to this soon please?. Again, very sorry to have needed to close the original pull because I really messed something up with git and couldn't figure it out.

**Comment by Garyfallidis** [comment]
Thank you @samcoveney I have assigned @ShreyasFadnavis to check your PR. You should hear from him asap. Thank you!

**Comment by skoudoro** [comment]
No worries for closing the previous one!

Apologize for the long wait, I tried to ping some expert in denoising but no feedback until now.

Overall, it looks good to me. I need to test a last time to refresh my memory since since it is was long time ago. 

@arokem, @ garyfallidis, any opinions?

**Comment by samcoveney** [comment]
Thanks everyone! Looking forward to comments

**Review by RafaelNH — CHANGES_REQUESTED** [review]
Hi @samcoveney! Thanks for your work on this! Only have 4 minor comments that I would like you to addressed. After that, I am happy that your PR is merged.

File: dipy/denoise/pca_noise_estimate.pyx
Comment (samcoveney) [reply dipy/denoise/pca_noise_estimate.pyx:65]: Thanks. As I said in my botched pull request #2707 I have tried to contact the authors but have gotten no reply.

File: dipy/denoise/localpca.py
Comment (samcoveney) [reply dipy/denoise/localpca.py:45]: Although I had to remind myself of this, I think it is fairly common knowledge that subtracting the mean from the data matrix will remove a degree of freedom, such that the number of non-zero eigenvalues that can be found is reduced by one.  As a quick example: https://stats.stackexchange.com/questions/123318/why-are-there-only-n-1-principal-components-for-n-data-if-the-number-of-dime I think the alternative is to not centre the data matrix, which appears to be the case in the original publications, but probably it's better to centre the data in general. So either we centre the data and make an attempt to correct for the lost degree of freedom, which differs from the original publication, or we centre the data and make no attempt to correct, which also differs from the original publication but seems inconsistent to me.

File: dipy/denoise/localpca.py
Comment (samcoveney) [reply dipy/denoise/localpca.py:42]: That's probably a better idea than putting it in the docs, I'll get around to doing that, thanks

File: dipy/denoise/localpca.py
Comment (RafaelNH) [reply dipy/denoise/localpca.py:45]: Yes I understand that. But it is not clear for me why this loss of degree of freedom manifests on parameter Gamma = N/M in the MP distribution (line 66).

Comment (RafaelNH) [reply dipy/denoise/localpca.py:45]: PS: I am referring to the parameter gamma in veraart's MP-PCA manuscript

Comment (RafaelNH) [reply dipy/denoise/localpca.py:45]: Reviewing the original MP paper by Marcenko and Pastur, MP distribution is defined for M, N -> inf so that gamma = N/M, so the discussion on this may be irrelevant. But, anyway, if you are not sure about it, I will recommend to leave it without removing anything to keep it consistent to the original article.

File: dipy/denoise/pca_noise_estimate.pyx
Comment (skoudoro) [reply dipy/denoise/pca_noise_estimate.pyx:53]: I confirm, just add it as `optional`,

no default value to simplify the maintenance. We always forget to change it in both places in case of change.

File: dipy/denoise/pca_noise_estimate.pyx
Comment (RafaelNH) [reply dipy/denoise/pca_noise_estimate.pyx:65]: Ok. So ignore this comment.

File: dipy/denoise/localpca.py
Comment (samcoveney) [reply dipy/denoise/localpca.py:45]: I think you are right @RafaelNH I'll return the value to the original (the test to first discard zero eigenvalues, which uses N - 1, should remain though).

**Comment by samcoveney** [comment]
I have addressed the comments / changes and squashed everything. Thanks!

**Review by RafaelNH — APPROVED** [review]

**Comment by RafaelNH** [comment]
Thanks @samcoveney ! I don't have anything to add.

**Comment by RafaelNH** [comment]
Hi @samcoveney!
Actually I have a last comment. Since you are changing already existing DIPY procedures, can you report your changes in the dipy [api_change documentation](https://github.com/dipy/dipy/blob/master/doc/api_changes.rst?plain=1#L8)? Please create a new section for DIPY 1.7.0 and add a description of your changes there. Many thanks!

**Comment by skoudoro** [comment]
Thank you @RafaelNH and @jhlegarreta for the review. 

Thank you @samcoveney for this PR.

I will wait until monday/tuesday to merge this PR, just in case there are any additional comments

**Comment by skoudoro** [comment]
Thank you for this work @samcoveney and sorry for the delay, I believed it was merged before the break....

**Pull request merged**
